lib/srv/db/elasticsearch/engine.go

```diff
@@ -0,0 +1,207 @@
+/*
+
+ Copyright 2022 Gravitational, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package elasticsearch
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/common/role"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func init() {
+	common.RegisterEngine(newEngine, defaults.ProtocolElasticsearch)
+}
+
+// newEngine create new elasticsearch engine.
+func newEngine(ec common.EngineConfig) common.Engine {
+	return &Engine{EngineConfig: ec}
+}
+
+type Engine struct {
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn net.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+}
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+	return nil
+}
+
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	// Assume internal server error HTTP 500 and override if possible.
+	statusCode := http.StatusInternalServerError
+	if trace.IsAccessDenied(err) {
+		statusCode = http.StatusUnauthorized
+	}
+
+	jsonBody, err := json.Marshal(struct {
+		Success bool   `json:"success"`
+		Message string `json:"message"`
+	}{
+		Success: false,
+		Message: err.Error(),
+	})
```

Comment:
Does this is native ElasticSearch error response format ?  I in docs they are using following format
```
{
    "error" : {
        "root_cause" : [
            {
                "type" : "resource_not_found_exception",
                "reason" : "the task with id [does_not_exist] doesn't exist",
                "stack_trace": ...
            }
        ],
        "type" : "resource_not_found_exception",
        "reason" : "the task with id [does_not_exist] doesn't exist",
        "stack_trace": ...
    },
    "status": 404
}
```
Ref
https://github.com/elastic/go-elasticsearch/blob/e50e5ce23d15356e4293c95fdd3e2cf8e15f671c/typedapi/types/errorcause.go#L32